### PR TITLE
#14 add member overview

### LINKED ISSUE #14: Client: Add a members overview
The members overview displays a table with all members (with following data: name, phone number, status, e-mail). 
This table can be filtered and sorted.

### PR DESCRIPTION
**Pull request by LennartRukower**
Added the member overview page, which displays a table with all members and filter/sort-options.
Closes #14

### CODE REVIEW

File: client/src/ts/members/MemberOverview.tsx [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+      headers: {Authorization : `Bearer ${localStorage.getItem("token")}`}
+    })
+    .then((res) => {
+      if (res.status === 200){
+        if(mounted) {
+          setMembers(res.data);
+        }
+      }
+    });
+
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:195]: .catch block is missing

File: client/src/ts/members/MemberOverview.tsx [hunk 2]
```diff
@@ -0,0 +1,531 @@
...
+    // Clean-up function
+    return () => {mounted = false;};
+  };
+
+  useEffect(() => getMembers(), []);
+
+  /**
+   * Handles the change event on the search filter input
+   *
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:204]: Remove new line

File: client/src/ts/members/MemberOverview.tsx [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+   * @param event
+   */
+  const handleSearchInput = (event: React.ChangeEvent<HTMLInputElement>): void => {
+    setSearchFilter(event.target.value);
+  };
+
+  /**
+   * Handles the change event on the status filter input
+   *
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:213]: Remove new line

File: client/src/ts/members/MemberOverview.tsx [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+   * @param event
+   */
+  const handleStatusChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
+    setStatusFilter(event.target.value);
+  };
+
+  /**
+   * Handles the change event on the ressort filter input
+   *
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:222]: Remove new line

File: client/src/ts/members/MemberOverview.tsx [hunk 5]
```diff
@@ -0,0 +1,531 @@
...
+   * @param event
+   */
+  const handleRessortChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
+    setRessortFilter(event.target.value);
+  };
+
+  /**
+   * Handles the change event on the sort input
+   *
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:231]: Remove new line

File: client/src/ts/members/MemberOverview.tsx [hunk 6]
```diff
@@ -0,0 +1,531 @@
...
+    setSortOption(event.target.value);
+  };
+
+  /**
+   * Filters and sorts the member data and returns it
+   */
+  const getFilteredAndSortedMembers = ():Member[] => {
+    let filteredMembers = members;
+
+    // Filters after status
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:244]: Filters by

File: client/src/ts/members/MemberOverview.tsx [hunk 7]
```diff
@@ -0,0 +1,531 @@
...
+    if (statusFilter !== "") {
+      filteredMembers = filteredMembers.filter(member => {
+        return member.mitgliedstatus.toString() === statusFilter;
+      });
+    }
+
+    // Filters after ressort
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:251]: Filters by

File: client/src/ts/members/MemberOverview.tsx [hunk 8]
```diff
@@ -0,0 +1,531 @@
...
+    if (ressortFilter !== "") {
+      filteredMembers = filteredMembers.filter(member => {
+        return member.ressort === ressortFilter;
+      });
+    }
+
+    // Filters after search input
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:258]: Filters by

File: client/src/ts/members/MemberOverview.tsx [hunk 9]
```diff
@@ -0,0 +1,531 @@
...
+      return (
+        member.vorname.toLowerCase().includes(searchFilter.toLowerCase()) ||
+        member.nachname.toLowerCase().includes(searchFilter.toLowerCase()) ||
+        member.handy.toLowerCase().includes(searchFilter.toLowerCase())
+      );
+    });
+
+    let sortedMembers = filteredMembers;
+
+    // Sorts after last changed in ascending order
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:269]: Sorts by

File: client/src/ts/members/MemberOverview.tsx [hunk 10]
```diff
@@ -0,0 +1,531 @@
...
+
+        if (dateA < dateB) {
+          return -1;
+        } else if (dateA > dateB) {
+          return 1;
+        } else {
+          return 0;
+        }
+      });
+    // Sorts after last changed in descending order
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:283]: Sorts by

File: client/src/ts/members/MemberOverview.tsx [hunk 11]
```diff
@@ -0,0 +1,531 @@
...
+          return 1;
+        } else if (dateA > dateB) {
+          return -1;
+        } else {
+          return 0;
+        }
+      });
+    }
+
+    // Sorts after lastname in ascending alphabetical order
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:299]: Sorts by

File: client/src/ts/members/MemberOverview.tsx [hunk 12]
```diff
@@ -0,0 +1,531 @@
...
+    if (nameSort === "up") {
+      sortedMembers = sortedMembers.sort((a,b) => {
+        return a.nachname.localeCompare(b.nachname);
+      });
+    // Sorts after lastname in descending alphabetical order
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:304]: New line above comment
Sorts by

File: client/src/ts/members/MemberOverview.tsx [hunk 13]
```diff
@@ -0,0 +1,531 @@
...
+    } else if (nameSort === "down") {
+      sortedMembers = sortedMembers.sort((a,b) => {
+        return -a.nachname.localeCompare(b.nachname);
+      });
+    }
+
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:310]: Remove new Line

File: client/src/ts/members/MemberOverview.tsx [hunk 14]
```diff
@@ -0,0 +1,531 @@
...
+            </TextField>
+          </Grid>
+        </Grid>
+          <IconButton aria-label="more filter options" className={classes.filterBtn} onClick={toggleFilters}>
+            {additionalFiltersState ? <RemoveCircle fontSize="inherit" /> : <AddCircle fontSize="inherit" />}
+          </IconButton>
+        </form>
+        {additionalFiltersState ? additionalFilters : null}
+      </Paper>
+      <Paper className={classes.amountOfEntries}>
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:469]: Information is not relevant enough to justify own field i would propose to remove it

File: client/src/ts/members/MemberOverview.tsx [hunk 15]
```diff
@@ -0,0 +1,531 @@
...
+              <MenuItem value={"F&R"}>F&R</MenuItem>
+              <MenuItem value={"FK"}>FK</MenuItem>
+              <MenuItem value={"MIT"}>MIT</MenuItem>
+              <MenuItem value={"MAR"}>MAR</MenuItem>
+              <MenuItem value={"IT"}>IT</MenuItem>
+            </TextField>
+          </Grid>
+        </Grid>
+          <IconButton aria-label="more filter options" className={classes.filterBtn} onClick={toggleFilters}>
+            {additionalFiltersState ? <RemoveCircle fontSize="inherit" /> : <AddCircle fontSize="inherit" />}
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:464]: Icon is misleading i would recommend to use a downward arrow like in a dropdown menu

File: client/src/ts/members/MemberOverview.tsx [hunk 16]
```diff
@@ -0,0 +1,531 @@
...
+                </TableCell>
+                <TableCell
+                  className={classes.tableHeadCell}
+                  >
+                  Ressort
+                </TableCell>
+              </TableRow>
+            </TableHead>
+            <TableBody>
+              {getFilteredAndSortedMembers().map((member, i) => (
```
Comment (christophtrybek) [inline client/src/ts/members/MemberOverview.tsx:510]: No single literal variable names use "index" instead

File: client/src/ts/members/MemberOverview.tsx (hunk 14 above)
Comment (LennartRukower) [reply client/src/ts/members/MemberOverview.tsx:469]: The amount of entries was sth. that was requested. It is helpful to know how many entries there are, after the members were filtered.

Comment (christophtrybek) [reply client/src/ts/members/MemberOverview.tsx:469]: In that case it should be moved into the panel where filtering is applied

File: client/src/ts/members/MemberOverview.tsx [hunk 17]
```diff
@@ -0,0 +1,527 @@
...
+      headers: {Authorization : `Bearer ${localStorage.getItem("token")}`}
+    })
+    .then((res) => {
+      if (res.status === 200){
+        if(mounted) {
+          setMembers(res.data);
+        }
+      }
+    }).catch((error) => {
+      console.log(error);
```
Comment (qnewald) [inline client/src/ts/members/MemberOverview.tsx:193]: Avoid logging to console.

Comment (LennartRukower) [reply client/src/ts/members/MemberOverview.tsx:193]: This is a temporary solution, until a Snackbar is implemented, which is displayed, when a request fails.

**Pull request merged**